distributed/tests/test_cancelled_state.py

```diff
@@ -1292,6 +1294,106 @@ def test_secede_cancelled_or_resumed_workerstate(
     assert ts not in ws.long_running
 
 
+@gen_cluster(client=True, nthreads=[("", 1), ("", 1)], timeout=2)
+async def test_secede_racing_cancellation_and_scheduling_on_other_worker(c, s, a, b):
+    wsA = s.workers[a.address]
+    block_secede = Event()
+    block_long_running = Event()
+    handled_long_running = Event()
+
+    def f(block_secede, block_long_running):
+        block_secede.wait()
+        distributed.secede()
+        block_long_running.wait()
+        return 123
+
+    # Instrument long-running handler
+    original_handler = s.stream_handlers["long-running"]
+
+    async def instrumented_handle_long_running(*args, **kwargs):
+        try:
+            return original_handler(*args, **kwargs)
+        finally:
+            await handled_long_running.set()
+
+    s.stream_handlers["long-running"] = instrumented_handle_long_running
+
+    # Submit task and wait until it executes on a
+    x = c.submit(
+        f,
+        block_secede,
+        block_long_running,
+        key="x",
+        workers=[a.address],
+    )
+    await wait_for_state("x", "executing", a)
+
+    with captured_logger("distributed.scheduler", logging.ERROR) as caplog:
+        with freeze_batched_send(a.batched_stream):
+            # Let x secede (and later succeed) without informing the scheduler
+            await block_secede.set()
+            await wait_for_state("x", "long-running", a)
+            assert not a.state.executing
+            assert a.state.long_running
+            await block_long_running.set()
+
+            await wait_for_state("x", "memory", a)
+
+            # Cancel x while the scheduler does not know that it seceded
+            x.release()
+            await async_poll_for(lambda: not s.tasks, timeout=5)
+            assert not wsA.processing
+            assert not wsA.long_running
```

Comment:
I suggest to assert what you are assuming here. I suspect there is a race condition between the above wait_for_state and the assert I'm suggesting

```suggestion
            # Cancel x while the scheduler does not know that it seceded
            assert wsA.processing
            assert not wsA.long_running
            x.release()
            await async_poll_for(lambda: not s.tasks, timeout=5)
            assert not wsA.processing
            assert not wsA.long_running
```

Reply:
This is generally a little brittle. If the worker task is already in memory, chances are that the scheduler was already informed. The event blocking logic may have to be slightly adjusted.
How about releasing on the client before the task is in memory but while it is executing?

Reply:
I'm a bit confused by this comment. Note that we froze the stream from a to the scheduler, so the scheduler should remain uninformed.

Reply:
appologies, I missed the frozen stream